modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -3611,15 +3602,26 @@ else if (!spi.failureDetectionTimeoutEnabled() && (e instanceof
                                 if (latencyCheck && log.isInfoEnabled())
                                     log.info("Latency check message has been written to socket: " + msg.id());
 
+                                boolean ping = msg instanceof TcpDiscoveryConnectionCheckMessage;
+
+                                long timeout = timeoutHelper.nextTimeoutChunk(spi.getSocketTimeout());
+
+                                // For the ping we take half of actual failure detection. Another half is the interval.
                                 spi.writeToSocket(newNextNode ? newNext : next,
                                     sock,
                                     out,
                                     msg,
-                                    timeoutHelper.nextTimeoutChunk(spi.getSocketTimeout()));
+                                    ping && spi.failureDetectionTimeoutEnabled() ? timeout / 2 : timeout
+                                );
+
+                                timeout = timeoutHelper.nextTimeoutChunk(ackTimeout0);
 
                                 long tsNanos0 = System.nanoTime();
 
-                                int res = spi.readReceipt(sock, timeoutHelper.nextTimeoutChunk(ackTimeout0));
+                                int res = spi.readReceipt(sock, ping && spi.failureDetectionTimeoutEnabled() ?
+                                    timeout / 2 : timeout);
+
+                                updateLastSentMessageTime();
```

Comment:
1. This would require us to check connection often. Like `connCheckInterval = spi.failureDetectionTimeout() / 4.`. Is it OK?

2. Didn't catch you. We do wait for rest of the rest in second case:

`timeout = timeoutHelper.nextTimeoutChunk(ackTimeout0);` - this is rest of the rest. Makes sense?